Add functions to retrieve temp password for own user or other user (as rodsAdmin)

PaulVanSchayck opened the pull request:
Our users login to our iRODS web application using federated SAML. The web application talk to iRODS as that user, by using the iRODS proxy user and a rodsAdmin account. To offer webdav access to our users, we're using the temporary password functionality to create a short-lived password. 

So far we were using the PHP API for this. But we're migrating to Django. So we needed the temp password functionality also in the Python.

This adds temp password functionality to the Python API. Both the function for setting your own temp password, or setting someone else temp password (as rodsAdmin). For the temp_password_for_user() I've also added a test.

trel commented on irods/password_obfuscation.py:
```diff
@@ -275,3 +275,11 @@ def obfuscate_new_password(new, old, signature):
+
+
+def create_temp_password(temp_hash, source_password):
+    password = (temp_hash + source_password).ljust(100, chr(0))
```
can you talk about this set of choices?    why 100?   why chr(0)?    is md5 sufficient in this case?

PaulVanSchayck replied:
This was based on the iRODS-php code. 
https://github.com/MaastrichtUniversity/irods-php/blame/master/src/RODSConn.class.php#L324

The earliest source I can find for that is here:
https://github.com/DICE-UNC/irods-php/blame/472dc46024e2122618a2d82daac03211396333a1/prods/src/RODSConn.class.php#L237

But I can also find it back here in Jargon:

https://github.com/DICE-UNC/jargon/blob/89b8b0bb86669b1e4dffc85b24e8f8abcc408b0c/jargon-core/src/main/java/org/irods/jargon/core/security/IRODSPasswordUtilities.java#L609

Isn't this simply how the iRODS server has implemented the passing of the temp password to a client? I presume to safely communicate the new secret password over a potential unencrypted connection. You pass just one part, and combine it with a secret the client has (the connection password).

The choice for MD5 is therefor also already made in the server. And yes it would be wise to change that to SHA. But I guess that's a protocol breaking change.

trel replied:
very good - thanks.

the temp password behavior is a corner i've not visited before.

PaulVanSchayck replied:
Yes.. it's not well known. Maybe a bit of a shame, they are useful in this federated SAML access situation where the user does not have a (known) password. But they could also be used as short-lives access tokens as a sort of API token.

d-w-moore commented:
@PaulVanSchayck , This looks good to me. I'll run these commits along with  of my changes for PRC compatibility with the impending release of iRODS 4.2.9 and then we can squash to a single commit. 

Also: I'm looking through to see if this fixes any identified issues, and the closest that I can come is #224 .  Anybody know?

trel commented:
Could be related... but I don't see a direct relationship with #224.

PaulVanSchayck commented:
It's related but different. 

I think using the temp password functionality may be a solution for them. Because in the `temp_password_for_user()` the password is hashed together with the password of the rodsAdmin account doing the request.

A similar situation as in #224 is when using the proxy user and `temp_password()`. This doesn't work (we tried). Because it tries to hash the password with the rodsAdmin user. And unlike in `temp_password_for_user()` that's not what is supposed to happen.

d-w-moore commented:
@PaulVanSchayck  Does your use case include allowing `rodsuser`s to change their own passwords without  a `rodsadmin`'s intervention?  If so, is it meaningful/useful to have a test around that as well?

trel commented:
@PaulVanSchayck Please squash this down to a single(?) commit and we'll get it merged.

Thanks!

PaulVanSchayck commented:
> @PaulVanSchayck Does your use case include allowing `rodsuser`s to change their own passwords without a `rodsadmin`'s intervention? If so, is it meaningful/useful to have a test around that as well?

Our use case doesn't. But this PR does have the code for that. There is indeed no test for that. I can add such a test. Where would you like it? In `login_auth_test.php`, `connection_test.py` or `user_group_test.py`? 

> @PaulVanSchayck Please squash this down to a single(?) commit and we'll get it merged.

Terrell, correct me if I'm wrong, don't you have the ability to squash the PR during a merge directly from the PR interface we are in now? With the options in the the "Merge PR" button. I wouldn't know how to do it otherwise, other than opening a new PR.

trel commented:
> don't you have the ability to squash the PR during a merge directly from the PR interface

Could be - will investigate.   I want to make sure you still get attribution if I'm the one who squashes/pushes.

trel commented:
> I wouldn't know how to do it otherwise, other than opening a new PR.

If you were to figure the rebase/squash... a force push to your existing branch (temp_password) would automatically update this PR with the new diff between that branch and the target (irods:master).

PaulVanSchayck commented:
@d-w-moore I've added a test for getting a temp password for your own account. It's in in `login_auth_test.py`, as it's probably most appropriate. But it's quite different from the tests already in there. Open for other suggestions. Maybe a whole separate test class for temp passwords may be nicest alternative.

@trel OK, I will force-push this to one commit. If you're also happy with this extra test I've added.

trel commented:
This latest test is still for the `rodsadmin` to get a temppassword for their own account.   I think Dan was asking about a non-admin `rodsuser` being able to use this facility.

PaulVanSchayck commented:
Right, indeed, it was still a rodsAdmin requesting a temp password for himself. Now changed to a true new rodsUser requesting the password in one session, then starting a new session with that password. I'm now creating a new user for this. Let me know if you prefer me reusing the `alissa` user of the class. But I think I would rather prefer a whole separate test class in that case.

trel commented:
Creating and removing `bobby` is fine.

@d-w-moore can you do a quick review and confirm we're ready for Paul to squash and force-push to a single commit?

d-w-moore commented:
Will do.

d-w-moore commented:
Paul, this looks great! I am going to follow your suggestion as this should be put in its own class.  Can we add this patch to avoid it being skipped for the default case (no `alissa` ssh test user)

```diff
From c45b04b9dd6c336886a575704c41af599512b4c7 Mon Sep 17 00:00:00 2001
From: d-w-moore <dmoore@renci.org>
Date: Mon, 26 Apr 2021 13:58:13 +0000
Subject: [PATCH] separate test class for test_temp_password

---
 irods/test/login_auth_test.py | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/irods/test/login_auth_test.py b/irods/test/login_auth_test.py
index 2b8726a..6d7a57b 100644
--- a/irods/test/login_auth_test.py
+++ b/irods/test/login_auth_test.py
@@ -325,6 +325,16 @@ class TestLogins(unittest.TestCase):
     def test_8(self):
         self.tst0 ( ssl_opt = False, auth_opt = 'pam'    , env_opt = True  )
 
+class TestTempPassword(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.admin = helpers.make_session()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.admin.cleanup()
+
     def test_temp_password(self):
         # Make a new user
         new_user = 'bobby'
-- 
2.7.4


```

d-w-moore commented:
I would say at this point, with the application of my patch, go ahead and squash and we'll merge. The tests pass for me.

PaulVanSchayck commented:
Ok, that works for me,

Alternatively, I can also put both tests (`test_temp_password` and `test_set_temp_password`) in their own separate `temp_password_test.py` file. That makes the most sense to me.

trel commented:
> in their own separate `temp_password_test.py` file

sure, go for it.    thanks!

PaulVanSchayck commented:
Ok done. Let me know if you see any remaining issues

d-w-moore reviewed (commented):
Looks good, and it seems ready for squashing and merging.

trel commented:
yes, Paul, please squash to single commit, force-push to your temp_password branch... and make sure you have the most best commit message/body ever :)

PaulVanSchayck commented:
All done

trel commented:
So pretty!   Merging!
